File: src/eyes/posture_tick_engine.py
"""PostureTickEngine — unified posture timing engine."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional, Union

from .classifier import PoseState
from .types import WarningLevel, WarningLevelEvent


@dataclass(frozen=True)
class CorrectionEvent:
    direction: PoseState
    dimension: str = "yaw"  # "yaw" 或 "pitch"


@dataclass(frozen=True)
class GoodPostureEvent:
    pass


@dataclass(frozen=True)
class EyeRestEvent:
    pass


SenseEvent = Union[CorrectionEvent, GoodPostureEvent, EyeRestEvent, WarningLevelEvent]

_DEFAULT_OFF_AXIS_STREAK_THRESHOLD = 0.3
_DEFAULT_OFF_AXIS_REPEAT_INTERVAL = 10.0
_DEFAULT_FACING_THRESHOLD = 300.0
_DEFAULT_EYEREST_THRESHOLD = 900.0


class PostureTickEngine:
    """Unified engine tracking off-axis streaks, facing time, presence time, and warning levels."""

    def __init__(
        self,
        *,
        off_axis_streak_threshold_seconds: float | None = None,
        off_axis_repeat_interval_seconds: float | None = None,
        facing_threshold_seconds: float | None = None,
        eyest_threshold_seconds: float | None = None,
    ) -> None:
        self._off_axis_streak_threshold = (
            off_axis_streak_threshold_seconds
            if off_axis_streak_threshold_seconds is not None
            else _DEFAULT_OFF_AXIS_STREAK_THRESHOLD
        )
        self._off_axis_repeat_interval = (
            off_axis_repeat_interval_seconds
            if off_axis_repeat_interval_seconds is not None
            else _DEFAULT_OFF_AXIS_REPEAT_INTERVAL
        )
        self._facing_threshold = (
            facing_threshold_seconds
            if facing_threshold_seconds is not None
            else _DEFAULT_FACING_THRESHOLD
        )
        self._eyest_threshold = (
            eyest_threshold_seconds
            if eyest_threshold_seconds is not None
            else _DEFAULT_EYEREST_THRESHOLD
        )
        # Yaw 轴 streak 追踪
        self._yaw_streak: float = 0.0
        self._yaw_repeat_due_at: Optional[float] = None
        self._yaw_last_emit_at: Optional[float] = None
        # Pitch 轴 streak 追踪
        self._pitch_streak: float = 0.0
        self._pitch_repeat_due_at: Optional[float] = None
        self._pitch_last_emit_at: Optional[float] = None
        self._facing_seconds: float = 0.0
        self._presence_seconds: float = 0.0
        self._snoozed: bool = False
        self._warning_level: WarningLevel = WarningLevel.NORMAL
        self._warning_direction: Optional[str] = None
        self._warning_event: Optional[WarningLevelEvent] = None
        self._off_axis_continuous_seconds: float = 0.0
        self._corrected_remaining_seconds: float = 0.0

    @property
    def is_snoozed(self) -> bool:
        return self._snoozed

    def snooze(self) -> None:
        self._snoozed = True

    def resume(self) -> None:
        self._snoozed = False

    def reconfigure(
        self,
        *,
        off_axis_streak_threshold_seconds: float | None = None,
        off_axis_repeat_interval_seconds: float | None = None,
        facing_threshold_seconds: float | None = None,
        eyest_threshold_seconds: float | None = None,
    ) -> None:
        """Update timing thresholds at runtime without recreating the engine."""
        if off_axis_streak_threshold_seconds is not None:
            self._off_axis_streak_threshold = off_axis_streak_threshold_seconds
        if off_axis_repeat_interval_seconds is not None:
            self._off_axis_repeat_interval = off_axis_repeat_interval_seconds
        if facing_threshold_seconds is not None:
            self._facing_threshold = facing_threshold_seconds
        if eyest_threshold_seconds is not None:
            self._eyest_threshold = eyest_threshold_seconds

    def tick(self, yaw_state: PoseState, pitch_state: PoseState, dt: float) -> list[SenseEvent]:
        """处理一个 tick，yaw 和 pitch 独立追踪 streak。

        正对屏幕 = 两个轴都在阈值内。
        有脸 = 任意轴不是 NO_FACE。
        Warning level 仅由 yaw 偏离驱动；pitch 偏离触发 CorrectionEvent 但不升级 warning。
        """
        events: list[SenseEvent] = []
        self._warning_event = None

        if self._snoozed:
            return events

        # --- Yaw 轴 streak 追踪 ---
        if yaw_state in (PoseState.OFF_AXIS_LEFT, PoseState.OFF_AXIS_RIGHT):
            self._yaw_streak += dt
            if self._yaw_streak >= self._off_axis_streak_threshold:
                if self._yaw_last_emit_at is None:
                    self._yaw_last_emit_at = self._yaw_streak
                    self._yaw_repeat_due_at = self._yaw_streak + self._off_axis_repeat_interval
                    events.append(CorrectionEvent(direction=yaw_state, dimension="yaw"))
                elif self._yaw_repeat_due_at is not None and self._yaw_streak >= self._yaw_repeat_due_at:
                    self._yaw_repeat_due_at = self._yaw_streak + self._off_axis_repeat_interval
                    events.append(CorrectionEvent(direction=yaw_state, dimension="yaw"))
        else:
            self._yaw_streak = 0.0
            self._yaw_repeat_due_at = None
            self._yaw_last_emit_at = None

        # --- Pitch 轴 streak 追踪 ---
        if pitch_state in (PoseState.HEAD_UP, PoseState.HEAD_DOWN):
            self._pitch_streak += dt
            if self._pitch_streak >= self._off_axis_streak_threshold:
                if self._pitch_last_emit_at is None:
                    self._pitch_last_emit_at = self._pitch_streak
                    self._pitch_repeat_due_at = self._pitch_streak + self._off_axis_repeat_interval
                    events.append(CorrectionEvent(direction=pitch_state, dimension="pitch"))
                elif self._pitch_repeat_due_at is not None and self._pitch_streak >= self._pitch_repeat_due_at:
                    self._pitch_repeat_due_at = self._pitch_streak + self._off_axis_repeat_interval
                    events.append(CorrectionEvent(direction=pitch_state, dimension="pitch"))
        else:
            self._pitch_streak = 0.0
            self._pitch_repeat_due_at = None
            self._pitch_last_emit_at = None

        # --- 正对屏幕时间累积（两个轴都在阈值内才算） ---
        is_facing = yaw_state == PoseState.FACING_SCREEN and pitch_state == PoseState.FACING_SCREEN
        if is_facing:
            self._facing_seconds += dt
            if self._facing_seconds >= self._facing_threshold:
                self._facing_seconds = 0.0
                events.append(GoodPostureEvent())

        # --- 有脸时间累积（任意轴不是 NO_FACE 即有脸） ---
        has_face = yaw_state != PoseState.NO_FACE
        if has_face:
            self._presence_seconds += dt
            if self._presence_seconds >= self._eyest_threshold:
                self._presence_seconds = 0.0
                events.append(EyeRestEvent())

        # --- Warning level 状态机（仅 yaw 驱动） ---
        if yaw_state in (PoseState.OFF_AXIS_LEFT, PoseState.OFF_AXIS_RIGHT):
            direction = "left" if yaw_state == PoseState.OFF_AXIS_LEFT else "right"
            if self._warning_level in (WarningLevel.NORMAL, WarningLevel.CORRECTED):
                self._warning_level = WarningLevel.WARNING
                self._warning_direction = direction
                self._off_axis_continuous_seconds = dt
                self._warning_event = WarningLevelEvent(
                    level=WarningLevel.WARNING,
                    direction=direction,
                )
                events.append(self._warning_event)
            else:
                self._warning_direction = direction
                self._off_axis_continuous_seconds += dt
                if (
                    self._warning_level == WarningLevel.WARNING
                    and self._off_axis_continuous_seconds >= self._off_axis_repeat_interval
                ):
                    self._warning_level = WarningLevel.SEVERE
                    self._warning_event = WarningLevelEvent(
                        level=WarningLevel.SEVERE,
                        direction=direction,
                    )
                    events.append(self._warning_event)
        elif is_facing:
            if self._warning_level in (WarningLevel.WARNING, WarningLevel.SEVERE):
                self._warning_level = WarningLevel.CORRECTED
                self._corrected_remaining_seconds = 2.0
                self._warning_event = WarningLevelEvent(
                    level=WarningLevel.CORRECTED,
                    direction=None,
                )
                events.append(self._warning_event)
                self._off_axis_continuous_seconds = 0.0
            elif self._warning_level == WarningLevel.CORRECTED:
                self._corrected_remaining_seconds -= dt
                if self._corrected_remaining_seconds <= 0:
                    self._warning_level = WarningLevel.NORMAL
                    self._warning_event = WarningLevelEvent(
                        level=WarningLevel.NORMAL,
                        direction=None,
                    )
                    events.append(self._warning_event)
                    self._corrected_remaining_seconds = 0.0
        elif yaw_state == PoseState.NO_FACE:
            if self._warning_level in (WarningLevel.WARNING, WarningLevel.SEVERE, WarningLevel.CORRECTED):
                self._warning_level = WarningLevel.NORMAL
                self._warning_event = WarningLevelEvent(
                    level=WarningLevel.NORMAL,
                    direction=None,
                )
                events.append(self._warning_event)
                self._off_axis_continuous_seconds = 0.0
                self._corrected_remaining_seconds = 0.0

        return events

File: src-tauri/src/domain/posture_tick_engine.rs
use super::classifier::PoseState;

#[derive(Debug, Clone, Copy, PartialEq, Eq, serde::Serialize)]
pub enum WarningLevel {
    Normal,
    Warning,
    Severe,
    Corrected,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum SenseEvent {
    Correction {
        direction: PoseState,
    },
    GoodPosture,
    EyeRest,
    WarningLevelChanged {
        level: WarningLevel,
        direction: Option<String>,
    },
}

const DEFAULT_OFF_AXIS_STREAK_THRESHOLD: f64 = 0.3;
const DEFAULT_OFF_AXIS_REPEAT_INTERVAL: f64 = 10.0;
const DEFAULT_FACING_THRESHOLD: f64 = 300.0;
const DEFAULT_EYEREST_THRESHOLD: f64 = 900.0;

fn is_yaw_state(state: PoseState) -> bool {
    matches!(state, PoseState::OffAxisLeft | PoseState::OffAxisRight)
}

fn is_pitch_state(state: PoseState) -> bool {
    matches!(state, PoseState::HeadUp | PoseState::HeadDown)
}

fn direction_label(state: PoseState) -> &'static str {
    match state {
        PoseState::OffAxisLeft => "left",
        PoseState::OffAxisRight => "right",
        PoseState::HeadUp => "up",
        PoseState::HeadDown => "down",
        _ => "",
    }
}

/// 单轴 off-axis 跟踪状态：streak 计时、重复提醒、警告升级 FSM。
#[derive(Debug, Clone)]
struct OffAxisState {
    streak: f64,
    repeat_due_at: Option<f64>,
    last_emit_at: Option<f64>,
    warning_level: WarningLevel,
    continuous_seconds: f64,
    corrected_remaining_seconds: f64,
}

impl OffAxisState {
    fn new() -> Self {
        Self {
            streak: 0.0,
            repeat_due_at: None,
            last_emit_at: None,
            warning_level: WarningLevel::Normal,
            continuous_seconds: 0.0,
            corrected_remaining_seconds: 0.0,
        }
    }

    fn reset_streak(&mut self) {
        self.streak = 0.0;
        self.repeat_due_at = None;
        self.last_emit_at = None;
    }

    fn reset_warning(&mut self) {
        self.warning_level = WarningLevel::Normal;
        self.continuous_seconds = 0.0;
        self.corrected_remaining_seconds = 0.0;
    }

    /// 偏离连续时长追踪。
    fn update_streak(
        &mut self,
        state: PoseState,
        dt: f64,
        streak_threshold: f64,
        repeat_interval: f64,
        events: &mut Vec<SenseEvent>,
    ) {
        if is_off_for_self(state) {
            self.streak += dt;
            if self.streak >= streak_threshold {
                if self.last_emit_at.is_none() {
                    self.last_emit_at = Some(self.streak);
                    self.repeat_due_at = Some(self.streak + repeat_interval);
                    events.push(SenseEvent::Correction { direction: state });
                } else if self
                    .repeat_due_at
                    .is_some_and(|due_at| self.streak >= due_at)
                {
                    self.repeat_due_at = Some(self.streak + repeat_interval);
                    events.push(SenseEvent::Correction { direction: state });
                }
            }
        } else {
            self.reset_streak();
        }
    }

    /// Warning 级别 FSM：Normal → Warning → Severe → Corrected → Normal。
    fn update_warning_level(
        &mut self,
        state: PoseState,
        dt: f64,
        repeat_interval: f64,
        events: &mut Vec<SenseEvent>,
    ) {
        let direction = direction_label(state);
        match state {
            _ if is_off_for_self(state) => {
                if matches!(
                    self.warning_level,
                    WarningLevel::Normal | WarningLevel::Corrected
                ) {
                    self.warning_level = WarningLevel::Warning;
                    self.continuous_seconds = dt;
                    events.push(SenseEvent::WarningLevelChanged {
                        level: WarningLevel::Warning,
                        direction: Some(direction.to_string()),
                    });
                } else {
                    self.continuous_seconds += dt;
                    if self.warning_level == WarningLevel::Warning
                        && self.continuous_seconds >= repeat_interval
                    {
                        self.warning_level = WarningLevel::Severe;
                        events.push(SenseEvent::WarningLevelChanged {
                            level: WarningLevel::Severe,
                            direction: Some(direction.to_string()),
                        });
                    }
                }
            }
            PoseState::FacingScreen => {
                if matches!(
                    self.warning_level,
                    WarningLevel::Warning | WarningLevel::Severe
                ) {
                    self.warning_level = WarningLevel::Corrected;
                    self.corrected_remaining_seconds = 2.0;
                    self.continuous_seconds = 0.0;
                    events.push(SenseEvent::WarningLevelChanged {
                        level: WarningLevel::Corrected,
                        direction: None,
                    });
                } else if self.warning_level == WarningLevel::Corrected {
                    self.corrected_remaining_seconds -= dt;
                    if self.corrected_remaining_seconds <= 0.0 {
                        self.warning_level = WarningLevel::Normal;
                        self.corrected_remaining_seconds = 0.0;
                        events.push(SenseEvent::WarningLevelChanged {
                            level: WarningLevel::Normal,
                            direction: None,
                        });
                    }
                }
            }
            PoseState::NoFace => {
                if matches!(
                    self.warning_level,
                    WarningLevel::Warning | WarningLevel::Severe | WarningLevel::Corrected
                ) {
                    self.reset_warning();
                    events.push(SenseEvent::WarningLevelChanged {
                        level: WarningLevel::Normal,
                        direction: None,
                    });
                }
            }
            _ => {}
        }
    }

    /// 处理单轴完整 tick：streak + warning。
    fn tick(
        &mut self,
        state: PoseState,
        dt: f64,
        streak_threshold: f64,
        repeat_interval: f64,
    ) -> Vec<SenseEvent> {
        let mut events = Vec::new();
        // 不属于本轴且非通用状态 → 静默忽略
        if !is_applicable(state) {
            return events;
        }
        self.update_streak(state, dt, streak_threshold, repeat_interval, &mut events);
        self.update_warning_level(state, dt, repeat_interval, &mut events);
        events
    }
}

/// 判断状态是否属于本轴的 off-axis。
fn is_off_for_self(state: PoseState) -> bool {
    // 调用前已通过 is_applicable 过滤，此处只需匹配所有 off-axis 状态
    matches!(
        state,
        PoseState::OffAxisLeft
            | PoseState::OffAxisRight
            | PoseState::HeadUp
            | PoseState::HeadDown
    )
}

/// 通用状态（两轴共享）：FacingScreen、NoFace。
fn is_applicable(state: PoseState) -> bool {
    is_yaw_state(state) || is_pitch_state(state) || matches!(state, PoseState::FacingScreen | PoseState::NoFace)
}

#[derive(Debug, Clone)]
pub struct PostureTickEngine {
    off_axis_streak_threshold: f64,
    off_axis_repeat_interval: f64,
    facing_threshold: f64,
    eyest_threshold: f64,
    facing_seconds: f64,
    presence_seconds: f64,
    snoozed: bool,
    yaw_oa: OffAxisState,
    pitch_oa: OffAxisState,
}

impl Default for PostureTickEngine {
    fn default() -> Self {
        Self::new(None, None, None, None)
    }
}

impl PostureTickEngine {
    pub fn new(
        off_axis_streak_threshold_seconds: Option<f64>,
        off_axis_repeat_interval_seconds: Option<f64>,
        facing_threshold_seconds: Option<f64>,
        eyest_threshold_seconds: Option<f64>,
    ) -> Self {
        Self {
            off_axis_streak_threshold: off_axis_streak_threshold_seconds
                .unwrap_or(DEFAULT_OFF_AXIS_STREAK_THRESHOLD),
            off_axis_repeat_interval: off_axis_repeat_interval_seconds
                .unwrap_or(DEFAULT_OFF_AXIS_REPEAT_INTERVAL),
            facing_threshold: facing_threshold_seconds.unwrap_or(DEFAULT_FACING_THRESHOLD),
            eyest_threshold: eyest_threshold_seconds.unwrap_or(DEFAULT_EYEREST_THRESHOLD),
            facing_seconds: 0.0,
            presence_seconds: 0.0,
            snoozed: false,
            yaw_oa: OffAxisState::new(),
            pitch_oa: OffAxisState::new(),
        }
    }

    pub fn is_snoozed(&self) -> bool {
        self.snoozed
    }

    pub fn snooze(&mut self) {
        self.snoozed = true;
    }

    pub fn resume(&mut self) {
        self.snoozed = false;
    }

    /// 当前综合警告级别（取两轴中更严重者）。
    pub fn warning_level(&self) -> WarningLevel {
        match (self.yaw_oa.warning_level, self.pitch_oa.warning_level) {
            (WarningLevel::Severe, _) | (_, WarningLevel::Severe) => WarningLevel::Severe,
            (WarningLevel::Warning, _) | (_, WarningLevel::Warning) => WarningLevel::Warning,
            (WarningLevel::Corrected, _) | (_, WarningLevel::Corrected) => WarningLevel::Corrected,
            _ => WarningLevel::Normal,
        }
    }

    /// 双轴 tick。
    ///
    /// - yaw_state / pitch_state 各自独立生成 Correction 和 WarningLevelChanged 事件。
    /// - GoodPosture 仅在两轴均 FacingScreen 时累积。
    /// - EyeRest 基于有脸时间（两轴均非 NoFace）。
    pub fn tick(
        &mut self,
        yaw_state: PoseState,
        pitch_state: PoseState,
        dt: f64,
    ) -> Vec<SenseEvent> {
        let mut events = Vec::new();

        if self.snoozed {
            return events;
        }

        // 各轴独立处理 off-axis streak + 警告升级
        events.extend(self.yaw_oa.tick(
            yaw_state,
            dt,
            self.off_axis_streak_threshold,
            self.off_axis_repeat_interval,
        ));
        events.extend(self.pitch_oa.tick(
            pitch_state,
            dt,
            self.off_axis_streak_threshold,
            self.off_axis_repeat_interval,
        ));

        // GoodPosture：两轴均 FacingScreen 才累积，其他情况暂停（不重置）
        let both_facing =
            yaw_state == PoseState::FacingScreen && pitch_state == PoseState::FacingScreen;
        if both_facing {
            self.facing_seconds += dt;
            if self.facing_seconds >= self.facing_threshold {
                self.facing_seconds = 0.0;
                events.push(SenseEvent::GoodPosture);
            }
        }

        // EyeRest：两轴均有脸即累积
        let any_face = yaw_state != PoseState::NoFace && pitch_state != PoseState::NoFace;
        if any_face {
            self.presence_seconds += dt;
            if self.presence_seconds >= self.eyest_threshold {
                self.presence_seconds = 0.0;
                events.push(SenseEvent::EyeRest);
            }
        }

        events
    }
}
